File: Anul II/Semestrul II/Medii de proiectare si programare/AgentiiZboruri/Persistence/src/main/java/ro/persistence/ZborDbRepo.java
package ro.persistence;

import ro.model.Entity;
import ro.model.Zbor;

import java.sql.Connection;
import java.sql.PreparedStatement;
import java.sql.ResultSet;
import java.sql.SQLException;
import java.time.LocalDateTime;
import java.time.format.DateTimeFormatter;
import java.util.*;

public class ZborDbRepo implements Repository{
    private final DbUtils dbUtils;

    private static final DateTimeFormatter formatter = DateTimeFormatter.ofPattern("yyyy-MM-dd HH:mm:ss");
    public ZborDbRepo(DbUtils dbUtils) {
        this.dbUtils = dbUtils;
    }

    @Override
    public Optional findOne(Object o) {
        Connection con = dbUtils.getConnection();
        try (PreparedStatement preStmt = con.prepareStatement("SELECT * FROM Zbor WHERE id=?")) {
            preStmt.setInt(1, (Integer) o);
            try (ResultSet result = preStmt.executeQuery()) {
                if (result.next()) {
                    Integer id=result.getInt("id");
                    String from=result.getString("from");
                    String to=result.getString("to");
                    String dataOra=result.getString("dataOra");
                    LocalDateTime dataOraDateTime = LocalDateTime.parse(dataOra, formatter);
                    Integer locuriDisponibile=result.getInt("locuriDisponibile");
                    Zbor zbor = new Zbor(id, from, to, dataOraDateTime, locuriDisponibile);
                    return Optional.of(zbor);
                }
            }
        } catch (SQLException ex) {
            System.out.println("Error DB " + ex);
        }
        return Optional.empty();
    }

    @Override
    public Iterable findAll() {
        Connection con = dbUtils.getConnection();
        List<Zbor> zboruri = new ArrayList<>();
        try (PreparedStatement preStmt = con.prepareStatement("SELECT * FROM Zbor")) {
            try (ResultSet result = preStmt.executeQuery()) {
                while (result.next()) {
                    Integer id=result.getInt("id");
                    String from=result.getString("from");
                    String to=result.getString("to");
                    String dataOra=result.getString("dataOra");
                    LocalDateTime dataOraDateTime = LocalDateTime.parse(dataOra, formatter);
                    Integer locuriDisponibile=result.getInt("locuriDisponibile");
                    Zbor zbor = new Zbor(id, from, to, dataOraDateTime, locuriDisponibile);
                    zboruri.add(zbor);
                }
            }
        } catch (SQLException ex) {
            System.out.println("Error DB " + ex);
        }
        return zboruri;
    }

    @Override
    public Optional save(Entity entity) {
        Connection con = dbUtils.getConnection();
        try (PreparedStatement preStmt = con.prepareStatement("INSERT INTO Zbor (id, from, to, dataOra, locuriDisponibile) VALUES (?, ?, ?, ?, ?)")) {
            Zbor zbor = (Zbor) entity;
            preStmt.setInt(1, zbor.getId());
            preStmt.setString(2, zbor.getFrom());
            preStmt.setString(3, zbor.getTo());
            String dataOraString = zbor.getDataOra().format(formatter);
            preStmt.setString(4, dataOraString);
            preStmt.setInt(5, zbor.getLocuriDisponibile());
            preStmt.executeUpdate();
        } catch (SQLException ex) {
            System.out.println("Error DB " + ex);
            return Optional.of(entity);
        }
        return Optional.empty();
    }

    public Iterable findAllFiltered(String to, String dataOra) {
        Connection con = dbUtils.getConnection();
        List<Zbor> zboruri = new ArrayList<>();
        try (PreparedStatement preStmt = con.prepareStatement("SELECT * FROM Zbor WHERE to = ? AND dataOra = ?")) {
            preStmt.setString(1, to);
            preStmt.setString(2, dataOra);
            try (ResultSet result = preStmt.executeQuery()) {
                while (result.next()) {
                    Integer id=result.getInt("id");
                    String from=result.getString("from");
                    String toResult=result.getString("to");
                    String dataOraR=result.getString("dataOra");
                    LocalDateTime dataOraDateTime = LocalDateTime.parse(dataOraR, formatter);
                    Integer locuriDisponibile=result.getInt("locuriDisponibile");
                    Zbor zbor = new Zbor(id, from, toResult, dataOraDateTime, locuriDisponibile);
                    zboruri.add(zbor);
                }
            }
        } catch (SQLException ex) {
            System.out.println("Error DB " + ex);
        }
        return zboruri;
    }

    @Override
    public Optional delete(Object o) {
        return Optional.empty();
    }

    @Override
    public Optional update(Entity entity) {
        return Optional.empty();
    }

    @Override
    public Iterable changeEntities(Map entities) {
        return null;
    }
}

File: Anul II/Semestrul II/Medii de proiectare si programare/AgentiiZboruriC/repository/ZborDbRepo.cs
using System.Collections;
using System.Data.SQLite;
using AgentiiZboruriC.domain;
using log4net.Repository.Hierarchy;
using System;
using System.Collections.Generic;
namespace AgentiiZboruriC.repository;
using System;
using System.Collections.Generic;
using System.Data.SqlClient;
using log4net;

public class ZborDbRepo : Repository<int, Zbor>
    {
        private readonly DbUtils dbUtils;

        public ZborDbRepo(DbUtils dbUtils)
        {
            this.dbUtils = dbUtils;
        }

        public Zbor? FindOne(object o)
        {
            using (SQLiteConnection con = dbUtils.GetConnection())
            using (SQLiteCommand cmd = new SQLiteCommand("SELECT * FROM Zbor WHERE id=@id", con))
            {
                cmd.Parameters.AddWithValue("@id", (int)o);
                using (SQLiteDataReader result = cmd.ExecuteReader())
                {
                    if (result.Read())
                    {
                        int id = result.GetInt32(0);
                        string from = result.GetString(1);
                        string to = result.GetString(2);
                        string dataOra = result.GetString(3);
                        DateTime dataOraDateTime = DateTime.Parse(dataOra);
                        int locuriDisponibile = result.GetInt32(4);
                        Zbor zbor = new Zbor(id, from, to, dataOraDateTime, locuriDisponibile);
                        return zbor;
                    }
                }
            }
            return null;
        }

        public Zbor FindOne(int id)
        {
            throw new NotImplementedException();
        }

        public IEnumerable<Zbor> FindAll()
        {
            List<Zbor> zboruri = new List<Zbor>();
            using (SQLiteConnection con = dbUtils.GetConnection())
            using (SQLiteCommand cmd = new SQLiteCommand("SELECT * FROM Zbor", con))
            {
                using (SQLiteDataReader result = cmd.ExecuteReader())
                {
                    while (result.Read())
                    {
                        int id = result.GetInt32(0);
                        string from = result.GetString(1);
                        string to = result.GetString(2);
                        string dataOra = result.GetString(3);
                        DateTime dataOraDateTime = DateTime.Parse(dataOra);
                        int locuriDisponibile = result.GetInt32(4);
                        Zbor zbor = new Zbor(id, from, to, dataOraDateTime, locuriDisponibile);
                        zboruri.Add(zbor);
                    }
                }
            }
            return zboruri;
        }

        Optional<Zbor> Repository<int, Zbor>.Save(Zbor entity)
        {
            throw new NotImplementedException();
        }

        public Optional<Zbor> Delete(int id)
        {
            throw new NotImplementedException();
        }

        Optional<Zbor> Repository<int, Zbor>.Update(Zbor entity)
        {
            throw new NotImplementedException();
        }

        public Zbor? Save(Zbor entity)
        {
            using (SQLiteConnection con = dbUtils.GetConnection())
            using (SQLiteCommand cmd = new SQLiteCommand("INSERT INTO Zbor (id, from, to, dataOra, locuriDisponibile) VALUES (@id, @from, @to, @dataOra, @locuriDisponibile)", con))
            {
                cmd.Parameters.AddWithValue("@id", entity.Id);
                cmd.Parameters.AddWithValue("@from", entity.From);
                cmd.Parameters.AddWithValue("@to", entity.To);
                cmd.Parameters.AddWithValue("@dataOra", entity.DataOra.ToString("yyyy-MM-dd HH:mm:ss"));
                cmd.Parameters.AddWithValue("@locuriDisponibile", entity.LocuriDisponibile);
                cmd.ExecuteNonQuery();
            }
            return null;
        }

        public IEnumerable<Zbor> FindAllFiltered(string to, string dataOra)
        {
            List<Zbor> zboruri = new List<Zbor>();
            using (SQLiteConnection con = dbUtils.GetConnection())
            using (SQLiteCommand cmd = new SQLiteCommand("SELECT * FROM Zbor WHERE to = @to AND dataOra = @dataOra", con))
            {
                cmd.Parameters.AddWithValue("@to", to);
                cmd.Parameters.AddWithValue("@dataOra", dataOra);
                using (SQLiteDataReader result = cmd.ExecuteReader())
                {
                    while (result.Read())
                    {
                        int id = result.GetInt32(0);
                        string from = result.GetString(1);
                        string toResult = result.GetString(2);
                        string dataOraR = result.GetString(3);
                        DateTime dataOraDateTime = DateTime.Parse(dataOraR);
                        int locuriDisponibile = result.GetInt32(4);
                        Zbor zbor = new Zbor(id, from, toResult, dataOraDateTime, locuriDisponibile);
                        zboruri.Add(zbor);
                    }
                }
            }
            return zboruri;
        }

        public Zbor? Delete(object o)
        {
            return null;
        }

        public Zbor? Update(Zbor entity)
        {
            return null;
        }

        public IEnumerable<Zbor> ChangeEntities(Dictionary<int, Zbor> entities)
        {
            return null;
        }
    }
